packages/agreement-process/src/model/domain/agreement-validators.ts

```diff
@@ -137,89 +139,94 @@ export const assertRequesterIsConsumer = (
 ): Promise<void> => {
+  // This operation has a dedicated assertion because it's the only operation that
+  // can be performed also by the producer even when there is an active producer delegation
   try {
     assertRequesterIsConsumer(agreement, authData);
   } catch (error) {
     try {
       assertRequesterIsProducer(agreement, authData);
     } catch (error) {
-      const producerDelegation =
-        await readModelService.getDelegationByDelegateId(
-          authData.organizationId,
-          delegationKind.delegatedProducer
+      const activeProducerDelegation =
+        await readModelService.getActiveProducerDelegationByEserviceId(
+          agreement.eserviceId
         );
-      assertRequesterIsDelegate(producerDelegation?.delegateId, authData);
+      assertRequesterIsDelegateProducer(
+        agreement,
+        authData,
+        activeProducerDelegation
+      );
     }
   }
 };
 
-export const assertRequesterIsProducerOrDelegateProducer = (
+export const assertRequesterCanActAsProducer = (
   agreement: Agreement,
-  delegateProducerId: TenantId | undefined,
-  authData: AuthData
+  authData: AuthData,
+  activeProducerDelegation: Delegation | undefined
 ): void => {
-  if (delegateProducerId) {
-    assertRequesterIsDelegate(delegateProducerId, authData);
-  } else {
+  if (!activeProducerDelegation) {
+    // No active producer delegation, requester is auhorized only if they are the producer
     assertRequesterIsProducer(agreement, authData);
+  } else {
+    // Active producer delegation, requester is authorized only if they are the delegate
+    assertRequesterIsDelegateProducer(
+      agreement,
+      authData,
+      activeProducerDelegation
+    );
   }
 };
 
-export const assertRequesterIsDelegate = (
-  delegateId: TenantId | undefined,
-  authData: AuthData
-): void => {
-  if (authData.organizationId !== delegateId) {
-    throw operationNotAllowed(authData.organizationId);
-  }
-};
-
-export const assertRequesterCanActivate = (
+const assertRequesterIsDelegateProducer = (
   agreement: Agreement,
-  delegateProducerId: TenantId | undefined,
-  authData: AuthData
+  authData: AuthData,
+  activeProducerDelegation: Delegation | undefined
 ): void => {
-  try {
-    assertRequesterIsConsumer(agreement, authData);
-  } catch (e) {
-    if (delegateProducerId) {
-      assertRequesterIsDelegate(delegateProducerId, authData);
-    } else {
-      assertRequesterIsProducer(agreement, authData);
-    }
+  if (
+    activeProducerDelegation?.delegateId !== authData.organizationId ||
+    activeProducerDelegation?.delegatorId !== agreement.producerId ||
+    activeProducerDelegation?.kind !== delegationKind.delegatedProducer ||
+    activeProducerDelegation?.state !== delegationState.active ||
+    activeProducerDelegation?.eserviceId !== agreement.eserviceId
+  ) {
+    throw operationNotAllowed(authData.organizationId);
   }
 };
 
-export const assertRequesterCanSuspend = assertRequesterCanActivate;
```

Comment:
The diff here is misleading: I modified only assertRequesterIsDelegate, while I removed assertRequesterCanActivate / assertRequesterCanSuspend because we can be more generic and use `assertRequesterCanActAsConsumerOrProducer` for both and we avoid defining a specific validator for each operation unless it's an exception like the one to retrieve consumer documents